File: day4.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)

void part1() {
	long long ans = 0;
	INPUT(line) {
		stringstream ss(line);
		string token;
		ss >> token >> token >> token;
		
		set<int> winners;
		while (token != "|") {
			winners.insert(stoi(token));
			ss >> token;
		}

		int value = 1;
		while (ss >> token)
			if (winners.count(stoi(token)))
				value *= 2;
		ans += value/2;
	}
	cout << ans;
}

void part2() {
	long long ans = 0;
	long long cards[1000];
	memset(cards, 0, sizeof cards);
	INPUT(line) {
		stringstream ss(line);
		string token;

		ss >> token >> token;
		token.pop_back();
		int game = stoi(token);
		ans += ++cards[game];

		ss >> token;
		set<int> winners;
		while (token != "|") {
			winners.insert(stoi(token));
			ss >> token;
		}

		int matches = 0;
		while (ss >> token)
			if (winners.count(stoi(token)))
				++matches;

		while (matches > 0) {
			cards[game+matches] += cards[game];
			--matches;
		}
	}
	cout << ans;
}

void main() {
	part1();
	part2();
}

File: day4.py
def part1():
    ans = 0
    while line := input():
        card, data = line.split(":")
        winners, given = data.split("|")
        winners = set(winners.split())

        val = 1
        for num in given.split():
            if num in winners:
                val *= 2
        ans += int(val/2)
    print(ans)

def part2():
    ans = 0
    cards = [1 for i in range(500)]
    while line := input():
        card, data = line.split(":")
        card = int(card[5:])
        winners, given = data.split("|")
        winners = set(winners.split())

        ans += cards[card]
        cnt = 0
        for num in given.split():
            if num in winners:
                cnt += 1
        for i in range(1, cnt+1):
            cards[card+i] += cards[card]
    print(ans)

part1()
part2()